packages/block-editor/src/hooks/custom-fields.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { PanelBody, TextControl } from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+import { hasBlockSupport } from '@wordpress/blocks';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { InspectorControls } from '../components';
+import { useBlockEditingMode } from '../components/block-editing-mode';
+
+/**
+ * Filters registered block settings, extending attributes to include `connections`.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Filtered block settings.
+ */
+export function addAttribute( settings ) {
+	if ( hasBlockSupport( settings, 'connections', true ) ) {
+		// Gracefully handle if settings.attributes is undefined.
+		settings.attributes = {
+			...settings.attributes,
+			connections: {
+				type: 'object',
+			},
+		};
+	}
+
+	return settings;
+}
+
+/**
+ * Override the default edit UI to include a new block inspector control for
+ * assigning a connection to blocks that has support for connections.
+ * Currently, only the `core/paragraph` block is supported and there is only a relation
+ * between paragraph content and a custom field.
+ *
+ * @param {WPComponent} BlockEdit Original component.
+ *
+ * @return {WPComponent} Wrapped component.
+ */
+export const withInspectorControl = createHigherOrderComponent(
+	( BlockEdit ) => {
+		return ( props ) => {
+			const blockEditingMode = useBlockEditingMode();
+			const hasCustomFieldsSupport = hasBlockSupport(
+				props.name,
+				'connections',
+				false
+			);
+			// We prevent that the content is lost when the user removes the custom field.
+			// Editing the content in the paragraph block with a placeholder is not the best solution.
+			const prevContent = useRef( props.attributes?.content );
+			if ( ! prevContent.current ) {
+				prevContent.current = '';
+			}
+			if ( hasCustomFieldsSupport && props.isSelected ) {
+				return (
+					<>
+						<BlockEdit { ...props } />
+						{ blockEditingMode === 'default' && (
+							<InspectorControls>
+								<PanelBody
+									title={ __( 'Connections' ) }
+									initialOpen={ true }
+								>
+									<TextControl
+										__nextHasNoMarginBottom
+										autoComplete="off"
+										label={ __( 'Custom field meta_key' ) }
+										value={
+											props.attributes?.connections
+												?.content?.value || ''
+										}
+										onChange={ ( nextValue ) => {
+											if ( nextValue === '' ) {
+												props.setAttributes( {
+													connections: undefined,
+													content:
+														prevContent.current !==
+														''
+															? prevContent.current
+															: undefined,
+												} );
+											} else {
+												props.setAttributes( {
+													connections: {
+														attributes: {
```

Comment:
I've added the intermediate `attributes` property here in 7575b404a65d93044bc96b3100e529ea1490d7a1

Reply:
Oh, we also needed 39cb9313bb5b429fba3852ad819e98903663b3ea !

Reply:
Just a note: We discussed this briefly and since we want to leave the door open for other types of connections (not just to attributes) it probably makes sense to have `connections.attributes` here. Another type of a connection might be to `style` in the parent block in the context of `wp:pattern` block. E.g.:

```js
{
    "connections": {
            "attributes": {
                "url": {
                    "source": "parent"
                }
            }
            "styles": {
                "color": {
                    "source": "parent"
                }
            }
    }
}
```